Unnecessary `Fn` bound for `SizeError::map_src`.

I am implementing a small wrapper over `read_from_bytes` that allows working over types like `Bytes` instead of `&[u8]`. As such I want to change the `Src` bound to be `Bytes` instead of `&[u8]`.

However, the bound on `map_src` is unnecessarily a `Fn(Src) -> NewSrc` which implies it would be callable multiple times. This is inconsistent with other forms of `map` which use `FnOnce`. This is a problem for me as I cannot simply `map_src(|_| bytes_buf)` as that moves ownership, giving me the error

```rust
error[E0507]: cannot move out of `buf`, a captured variable in an `Fn` closure
   --> proxy/src/protocol2.rs:371:40
    |
363 |         let buf = if self.len() < len {
    |             --- captured outer variable
...
371 |         res.map_err(|e| e.map_src(|()| buf))
    |                                   ---- ^^^ move occurs because `buf` has type `bytes::BytesMut`, which does not implement the `Copy` trait
    |                                   |
    |                                   captured by this `Fn` closure
    |
help: consider cloning the value if the performance cost is acceptable
    |
371 |         res.map_err(|e| e.map_src(|()| buf.clone()))
    |                                           ++++++++
```

Easy to fix here with a clone, sure. But it's not ideal.

I have a feeling this can be fixed without a breaking change, but I'm not 100% certain.

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -15,7 +15,7 @@
 [package]
 edition = "2021"
 name = "zerocopy"
-version = "0.8.8"
+version = "0.8.9"
 authors = ["Joshua Liebow-Feeser <joshlf@google.com>"]
 description = "Zerocopy makes zero-cost memory manipulation effortless. We write \"unsafe\" so you don't have to."
 categories = ["embedded", "encoding", "no-std::no-alloc", "parsing", "rust-patterns"]
@@ -81,13 +81,13 @@ std = ["alloc"]
 __internal_use_only_features_that_work_on_stable = ["alloc", "derive", "simd", "std"]
 
 [dependencies]
-zerocopy-derive = { version = "=0.8.8", path = "zerocopy-derive", optional = true }
+zerocopy-derive = { version = "=0.8.9", path = "zerocopy-derive", optional = true }
 
 # The "associated proc macro pattern" ensures that the versions of zerocopy and
 # zerocopy-derive remain equal, even if the 'derive' feature isn't used.
 # See: https://github.com/matklad/macro-dep-test
 [target.'cfg(any())'.dependencies]
-zerocopy-derive = { version = "=0.8.8", path = "zerocopy-derive" }
+zerocopy-derive = { version = "=0.8.9", path = "zerocopy-derive" }
 
 [dev-dependencies]
 itertools = "0.11"
@@ -101,6 +101,6 @@ testutil = { path = "testutil" }
 # CI test failures.
 trybuild = { version = "=1.0.89", features = ["diff"] }
 # In tests, unlike in production, zerocopy-derive is not optional
-zerocopy-derive = { version = "=0.8.8", path = "zerocopy-derive" }
+zerocopy-derive = { version = "=0.8.9", path = "zerocopy-derive" }
 # TODO(#381) Remove this dependency once we have our own layout gadgets.
 elain = "0.3.0"

--- a/src/error.rs
+++ b/src/error.rs
@@ -299,7 +299,7 @@ impl<Src, Dst: ?Sized> AlignmentError<Src, Dst> {
     /// });
     /// ```
     #[inline]
-    pub fn map_src<NewSrc>(self, f: impl Fn(Src) -> NewSrc) -> AlignmentError<NewSrc, Dst> {
+    pub fn map_src<NewSrc>(self, f: impl FnOnce(Src) -> NewSrc) -> AlignmentError<NewSrc, Dst> {
         AlignmentError { src: f(self.src), dst: SendSyncPhantomData::default() }
     }
 
@@ -455,7 +455,7 @@ impl<Src, Dst: ?Sized> SizeError<Src, Dst> {
     /// });
     /// ```
     #[inline]
-    pub fn map_src<NewSrc>(self, f: impl Fn(Src) -> NewSrc) -> SizeError<NewSrc, Dst> {
+    pub fn map_src<NewSrc>(self, f: impl FnOnce(Src) -> NewSrc) -> SizeError<NewSrc, Dst> {
         SizeError { src: f(self.src), dst: SendSyncPhantomData::default() }
     }
 
@@ -593,7 +593,7 @@ impl<Src, Dst: ?Sized + TryFromBytes> ValidityError<Src, Dst> {
     /// });
     /// ```
     #[inline]
-    pub fn map_src<NewSrc>(self, f: impl Fn(Src) -> NewSrc) -> ValidityError<NewSrc, Dst> {
+    pub fn map_src<NewSrc>(self, f: impl FnOnce(Src) -> NewSrc) -> ValidityError<NewSrc, Dst> {
         ValidityError { src: f(self.src), dst: SendSyncPhantomData::default() }
     }
 
@@ -714,7 +714,7 @@ impl<Src, Dst: ?Sized> CastError<Src, Dst> {
     /// });
     /// ```
     #[inline]
-    pub fn map_src<NewSrc>(self, f: impl Fn(Src) -> NewSrc) -> CastError<NewSrc, Dst> {
+    pub fn map_src<NewSrc>(self, f: impl FnOnce(Src) -> NewSrc) -> CastError<NewSrc, Dst> {
         match self {
             Self::Alignment(e) => CastError::Alignment(e.map_src(f)),
             Self::Size(e) => CastError::Size(e.map_src(f)),
@@ -835,7 +835,7 @@ impl<Src, Dst: ?Sized + TryFromBytes> TryCastError<Src, Dst> {
     ///     });
     /// ```
     #[inline]
-    pub fn map_src<NewSrc>(self, f: impl Fn(Src) -> NewSrc) -> TryCastError<NewSrc, Dst> {
+    pub fn map_src<NewSrc>(self, f: impl FnOnce(Src) -> NewSrc) -> TryCastError<NewSrc, Dst> {
         match self {
             Self::Alignment(e) => TryCastError::Alignment(e.map_src(f)),
             Self::Size(e) => TryCastError::Size(e.map_src(f)),
@@ -900,7 +900,7 @@ impl<Src, Dst: ?Sized + TryFromBytes> TryReadError<Src, Dst> {
     ///     });
     /// ```
     #[inline]
-    pub fn map_src<NewSrc>(self, f: impl Fn(Src) -> NewSrc) -> TryReadError<NewSrc, Dst> {
+    pub fn map_src<NewSrc>(self, f: impl FnOnce(Src) -> NewSrc) -> TryReadError<NewSrc, Dst> {
         match self {
             Self::Alignment(i) => match i {},
             Self::Size(e) => TryReadError::Size(e.map_src(f)),

--- a/zerocopy-derive/Cargo.toml
+++ b/zerocopy-derive/Cargo.toml
@@ -9,7 +9,7 @@
 [package]
 edition = "2021"
 name = "zerocopy-derive"
-version = "0.8.8"
+version = "0.8.9"
 authors = ["Joshua Liebow-Feeser <joshlf@google.com>"]
 description = "Custom derive for traits from the zerocopy crate"
 license = "BSD-2-Clause OR Apache-2.0 OR MIT"

